src/flint/types/fmpz_mpoly.pyx

```diff
@@ -135,362 +192,834 @@ cdef class fmpz_mpoly(flint_mpoly):
+            fmpz_mpoly u
+
         fmpz_mpoly_factor_init(fac, self.ctx.val)
-        fmpz_mpoly_factor(fac, self.val, 1, self.ctx.val)
-        res = [0] * fac.length
-        for 0 <= i < fac.length:
-            u = fmpz_mpoly.__new__(fmpz_mpoly)
-            u.ctx = self.ctx
+        fmpz_mpoly_factor_squarefree(fac, self.val, self.ctx.val)
+        res = [0] * fac.num
+
+        for i in range(fac.num):
+            u = create_fmpz_mpoly(self.ctx)
             fmpz_mpoly_init(u.val, u.ctx.val)
-            u._init = True
             fmpz_mpoly_set((<fmpz_mpoly>u).val, &fac.poly[i], self.ctx.val)
+
             c = fmpz.__new__(fmpz)
             fmpz_set((<fmpz>c).val, &fac.exp[i])
+
             res[i] = (u, c)
+
         c = fmpz.__new__(fmpz)
-        fmpz_set((<fmpz>c).val, fac.content)   # should be & with ...
+        fmpz_set((<fmpz>c).val, fac.constant)
         fmpz_mpoly_factor_clear(fac, self.ctx.val)
         return c, res
-    '''
 
+    # TODO: Rethink context conversions, particularly the proposed methods in #132
+    # def coerce_to_context(self, ctx):
+    #     cdef:
+    #         fmpz_mpoly res
+    #         slong *C
+    #         slong i
+
+    #     if not typecheck(ctx, fmpz_mpoly_ctx):
+    #         raise ValueError("provided context is not a fmpz_mpoly_ctx")
+
+    #     if self.ctx is ctx:
+    #         return self
+
+    #     C = <slong *> libc.stdlib.malloc(self.ctx.val.minfo.nvars * sizeof(slong *))
+    #     if C is NULL:
+    #         raise MemoryError("malloc returned a null pointer")
+    #     res = create_fmpz_mpoly(self.ctx)
+
+    #     vars = {x: i for i, x in enumerate(ctx.py_names)}
+    #     for i, var in enumerate(self.ctx.py_names):
+    #         C[i] = <slong>vars[var]
+
+    #     fmpz_mpoly_compose_fmpz_mpoly_gen(res.val, self.val, C, self.ctx.val, (<fmpz_mpoly_ctx>ctx).val)
+
+    #     libc.stdlib.free(C)
+    #     return res
+
+    def derivative(self, var):
+        """
+        Return the derivative of this polynomial with respect to the provided variable.
+        The argument can either be the variable as a string, or the index of the
+        variable in the context.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpz_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> p = ctx.from_dict({(0, 3): 2, (2, 1): 3})
+            >>> p
+            3*x0^2*x1 + 2*x1^3
+            >>> p.derivative("x0")
+            6*x0*x1
+            >>> p.derivative(1)
+            3*x0^2 + 6*x1^2
+
+        """
+        cdef:
+            fmpz_mpoly res
+            slong i = self.ctx.variable_to_index(var)
+
+        res = create_fmpz_mpoly(self.ctx)
+
+        fmpz_mpoly_derivative(res.val, self.val, i, self.ctx.val)
+        return res
+
+    def integral(self, var):
+        """
+        Return the integral of this polynomial*B with respect to the provided variable
+        where B is minimal. The argument and either be the variable as a string, or
```

Comment:
I don't understand what `polynomial*B` means here. Also and -> can

Reply:
This comes from the flint fmpz_mpoly integral method
>  ```
> void fmpz_mpoly_integral(fmpz_mpoly_t A, fmpz_t scale, const fmpz_mpoly_t B, slong var, const fmpz_mpoly_ctx_t ctx)
> ```
> Set A and scale so that A is an integral of `scale * B` with respect to the variable of index var, where scale is positive and as small as possible.

https://flintlib.org/doc/fmpz_mpoly.html#c.fmpz_mpoly_integral

Instead of producing a domain error when the polynomial isn't intergrable in Z, it scales the polynomial such that it would be, and spits out the scale in addition to the result.

Reply:
I see so the wording can be made clearer. Given p this method returns `(B, P)` such that `P' = B*p` or `P/B` is the formal integral of `p`.